robottelo/vm.py

```diff
@@ -214,37 +218,64 @@ def create(self):
         )
 
         result = ssh.command(
-                command,
-                self.provisioning_server,
-                connection_timeout=30
+            command,
+            self.provisioning_server,
+            connection_timeout=30
         )
-
         if result.return_code != 0:
             raise VirtualMachineError(
                 u'Failed to run snap-guest: {0}'.format(result.stderr))
         else:
             self._created = True
-
+            self.mac = [n.split('MAC:')[1].strip() for n in result.stdout if 'MAC:' in n][0]
         # outside of VLANs ping from hypervisor, in VLANs ping from SAT
         if self.bridge == 'br0':
             ping_from_hostname = self.provisioning_server
         else:
             ping_from_hostname = settings.server.hostname
 
         # Give some time to machine boot
-        result = ssh.command(
-            u'for i in {{1..60}}; do ping -c1 {0}.local && exit 0; sleep 1;'
-            u' done; exit 1'.format(self._target_image),
-            ping_from_hostname,
-            connection_timeout=30
-        )
-        if result.return_code != 0:
-            logger.error('Failed to obtain VM IP, reverting changes')
-            self.destroy()
-            raise VirtualMachineError(
-                'Failed to fetch virtual machine IP address information')
-        output = ''.join(result.stdout)
-        self.ip_addr = output.split('(')[1].split(')')[0]
+        for i in range(60):
+            qemu_ga_check = ssh.command(
+                u'virsh qemu-agent-command {0} '
+                '\'{{"execute":"guest-network-get-interfaces"}}\''.format(self.hostname),
+                ping_from_hostname,
+                connection_timeout=30
+            )
+            if qemu_ga_check.return_code != 0:
+                if 'guest agent is not connected' in qemu_ga_check.stderr:
+                    # this means the agent wasn't started yet (vm still booting)
+                    sleep(1)
+                else:
+                    # this means there's another error with agent, e.g. it is not configured
+                    break
+            else:
+                ifaces = json.loads(qemu_ga_check.stdout[0])
+                mgmt_if = next((i for i in ifaces['return']
+                               if i['hardware-address'].lower() == self.mac.lower()), {})
+                try:
+                    # get only the ipv4 addresses
+                    self.ip_addr = next(i['ip-address'] for i in mgmt_if['ip-addresses']
+                                        if i['ip-address-type'] == 'ipv4')
+                    break
+                except(KeyError, StopIteration):
+                    continue
+        if not self.ip_addr:
```

Comment:
can you pls add one empty line above to indicate end of the for loop

Reply:
I miss some sleep when ```qemu_ga_check``` command suceeds but there is still no IPv4 adress due to network start-up.

Reply:
It's handled by your previous (but chronologically later) comment. If there is no IPv4 address yet, generator will be empty and `next` will throw `StopIteration` on us.

Reply:
yes immediately